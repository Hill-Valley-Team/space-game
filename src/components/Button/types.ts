export type ButtonTypes = "submit" | "button" | "reset"

export type ButtonProps = {
  text?: string;
  className?: string;
  type?: ButtonTypes
};
