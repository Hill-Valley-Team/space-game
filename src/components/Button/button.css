.button {
  color: $text-color;
  border-radius: 50px;
  padding: 10px;
  border: none;
  cursor: pointer;

  &_width_stretch {
    width: 100%;
  }

  &_width_fixed {
    width: 280px;
  }

  &_align_center {
    margin: 0 auto;
  }

  &_align_right {
    right: 0;
  }

  &_view_primary {
    background-color: $primary-color;
    color: #fff;
  }

  &_view_default {
    background-color: $primary-bg-color;
  }

  &_view_link {
    background-color: transparent;
    color: $primary-color;
    text-decoration: underline;
    display: block;
    align-items: center;
    padding: 0;
    border-radius: unset;
  }

  &_view_secondary {
    border: 1px solid $primary-color;
    background: $secondary-btn-bg-color;
    color: $primary-color;
    display: flex;
    justify-content: center;
  }

  &_view_info {
    border: 1px solid $primary-color;
    background: transparent;
    color: $primary-color;
    border-radius: 8px;
    padding: 10px 20px;
    width: auto;
  }

  &_view_error {
    background: transparent;
    border: 1px solid $accent-color;
    color: $accent-color;
  }

  &:disabled {
    background-color: $disabled-bg-color;
    transform: none !important;
    translate: none !important;
    cursor: default !important;
  }

  &__link {
    text-decoration: none;
  }
}
