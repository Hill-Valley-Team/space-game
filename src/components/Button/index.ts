export { Button } from './Button';
<<<<<<< HEAD
export { ButtonProps, ButtonOwnProps } from './types';
=======
export { ButtonProps } from './types';
>>>>>>> sprint_5
