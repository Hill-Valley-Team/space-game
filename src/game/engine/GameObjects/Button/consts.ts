export const DEFAULT_BTN_COLOR = 'black';
export const DEFAULT_BTN_BG_COLOR = 'gray';
