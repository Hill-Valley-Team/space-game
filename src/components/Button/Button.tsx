import block from 'bem-cn';
import React from 'react';
import { ButtonProps } from './types';
import './button.css';

const b = block('button');

export const Button = ({
  text = 'Button',
  view = 'default',
  width = 'auto',
  align = 'left',
  className,
  type = 'submit',
  ...props
}: ButtonProps) => (
  <button type={type} className={b({ view, width, align }).mix(className)} {...props}>
    {text}
  </button>
);
