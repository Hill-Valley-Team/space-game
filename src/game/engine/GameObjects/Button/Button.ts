import { Mouse } from '../../Inputs';
import { GameObject } from '../GameObject/GameObject';
import { ButtonProps } from '../GameObject/types';
import { DEFAULT_BTN_BG_COLOR, DEFAULT_BTN_COLOR } from './consts';

export class Button extends GameObject {
  private text: string;

  private color: string;

  private bgColor: string;

  protected onClick: Function;

  public isClicked: boolean;

  constructor(props: ButtonProps) {
    const { scene, x, y, key, width, height, text, color, bgColor, onClick } = props;

    super({ scene, x, y, key, width, height });

    this.text = text ?? '';
    this.color = color ?? DEFAULT_BTN_COLOR;
    this.bgColor = bgColor ?? DEFAULT_BTN_BG_COLOR;
    this.isClicked = false;
    this.onClick = onClick ?? this.onBtnClick;
  }

  init() {
    this.addToScene();
    this.addListeners();
  }

  onBtnClick() {}

  checkClicked(event: MouseEvent) {
    this.scene.game.collider.clickCollide(event, this, this.onClick);
  }

  addListeners() {
    const self = this;
    this.scene.events.addListener(self, Mouse.NAMES.CLICK, this.checkClicked);
  }

  onRender() {
    const ctx = this.scene.game.context;

    if (ctx) {
      ctx.font = '20px Inter';
      ctx.fillStyle = this.bgColor;
      ctx.fillRect(this.x, this.y, this.width, this.height);

      ctx.fillStyle = this.color;
      ctx.fillText(
        this.text,
        this.x + this.width / 2 - ctx.measureText(this.text).width / 2,
        this.y + this.height / 2 + 5,
      );
    }
  }

  update() {
    if (this.isClicked && this.onClick) {
      this.onClick();
      this.isClicked = true;
    }
  }
}
