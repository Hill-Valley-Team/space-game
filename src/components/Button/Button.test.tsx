import React from 'react';
import renderer from 'react-test-renderer';
import { Button } from './Button';

describe('Button', () => {
  it('render with default props', () => {
    const tree = renderer.create(<Button text="Button" />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
