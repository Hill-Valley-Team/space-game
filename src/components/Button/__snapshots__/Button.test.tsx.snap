// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button render with default props 1`] = `
<button
  className={
    Object {
      "has": [Function],
      "is": [Function],
      "mix": [Function],
      "split": [Function],
      "state": [Function],
      "toString": [Function],
    }
  }
  type="submit"
>
  Button
</button>
`;
